libopenage/pathfinding/cost_field.h

```diff
@@ -95,15 +99,33 @@ class CostField {
 	 * Set the cost field values.
 	 *
 	 * @param cells Cost field values.
+	 * @param changed Time at which the cost value is changed.
 	 */
-	void set_costs(std::vector<cost_t> &&cells);
+	void set_costs(std::vector<cost_t> &&cells, const time::time_t &changed);
+
+	/**
+	 * Check if the cost field is dirty at the specified time.
+	 *
+	 * @param time Cost field is dirty if the cost field is accessed after the time given in valid_until.
+	 */
+	bool is_dirty(const time::time_t &time);
+
+	/**
+	 * Cleans the dirty flag by setting it to time_MAX.
+	 */
+	void clean();
 
 private:
 	/**
 	 * Side length of the field.
 	 */
 	size_t size;
 
+	/**
+	 * Time the cost field was last changed.
+	 */
+	time::time_t &valid_until;
```

Comment:
This member should not be a reference

```suggestion
	time::time_t valid_until;
```